app/internal/features.py

```diff
@@ -0,0 +1,187 @@
+    is_exists = session.query(
+        exists()
+        .where(Feature.name == feature["name"])
+        .where(Feature.route == feature["route"]),
+    ).scalar()
+
+    return is_exists
+
+
+def update_feature(
+    feature: Feature,
+    feature_dict: Dict[str, str],
+    session: Session = Depends(get_db),
+) -> Feature:
+    feature.name = feature_dict["name"]
+    feature.route = feature_dict["route"]
+    feature.description = feature_dict["description"]
+    feature.creator = feature_dict["creator"]
+    session.commit()
+    return feature
+
+
+async def is_access_allowd(request: Request, route: str) -> bool:
+    session = SessionLocal()
+
+    # Get current user.
+    # Note: can't use dependency beacause its designed for routes only.
+    # current_user return schema not an db model.
+    jwt = await get_authorization_cookie(request=request)
+    user = await current_user(request=request, jwt=jwt, db=session)
+
+    feature = session.query(Feature).filter_by(route=route).first()
+
+    if feature is None:
+        # in case there is no feature exists in the database that match the
+        # route that gived by to the request.
+        return True
+
+    user_ptef = session.query(
+        exists().where(
+            UserFeature.feature_id == feature.id
+        ).where(
+            UserFeature.user_id == user.user_id
+        )
+    ).scalar()
+
+    return user_ptef
+
+
+def create_feature(
+    name: str,
+    route: str,
+    description: str,
+    db: Session,
+    creator: str = None,
+) -> Feature:
+    """Creates a feature."""
+    return create_model(
+        db,
+        Feature,
+        name=name,
+        route=route,
+        creator=creator,
+        description=description,
+    )
+
+
+def create_user_feature_association(
+    db: Session,
+    feature_id: int,
+    user_id: int,
+    is_enable: bool,
+) -> UserFeature:
+    """Creates an association."""
+    return create_model(
+        db,
+        UserFeature,
+        user_id=user_id,
+        feature_id=feature_id,
+        is_enable=is_enable,
+    )
+
+
+def get_user_installed_features(
+    user_id: int,
+    session: Session = Depends(get_db),
+) -> List[Feature]:
+    return session.query(UserFeature).filter_by(user_id=user_id).all()
+
+
+async def get_user_uninstalled_features(
+    request: Request,
+    session: Session = Depends(get_db),
+) -> List[Feature]:
+    uninstalled = []
+    all_features = session.query(Feature).all()
+
+    # Get current user.
+    # Note: can't use dependency beacause its designed for routes only.
+    # current_user return schema not an db model.
+    jwt = await get_authorization_cookie(request=request)
+    user = await current_user(request=request, jwt=jwt, db=session)
+
+    for feat in all_features:
+        in_enabled = is_user_has_feature(
+            session=session,
+            feature_id=feat.id,
+            user_id=user.user_id,
+        )
```

Comment:
Create one unified query. It's important to understand that each query takes LOTS of time to run.
You should reduce it to one query.

Reply:
Did that!
Thanks for explaining.